### PR TITLE
Use latest .NET Core SDK

### PR DESCRIPTION
**Pull request by dougbu**
- move from dotnet-dev-1.0.0-preview2.1-003177 to dotnet-dev-1.0.4
- convert .NET Core Benchmarks project from project.json to .csproj
  - use latest available versions of all dependencies (1.1.3 patch train)
- get Windows scripts working
  - add appsettings.json file for Windows runs

nits:
- remove unused tool dependencies
- add and correct links in README.md
- remove duplicate property setting from benchmark_config.json

### CODE REVIEW

**Comment by dougbu** [comment]
Will not merge here. Instead will go into the TechEmpower repo.

File: frameworks/CSharp/aspnetcore/Benchmarks/Benchmarks.csproj [hunk 1]
```diff
@@ -0,0 +1,31 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+  <PropertyGroup>
+    <TargetFramework>netcoreapp1.1</TargetFramework>
+    <OutputType>Exe</OutputType>
+    <!-- NETStandard.Library version -->
+    <NetStandardImplicitPackageVersion>1.6.1</NetStandardImplicitPackageVersion>
```
Comment (mikeharder) [inline frameworks/CSharp/aspnetcore/Benchmarks/Benchmarks.csproj]: I would remove `NetStandardImplicitPackageVersion` and `RuntimeFrameworkVersion` unless most customer apps would set them.

File: frameworks/CSharp/aspnetcore/Benchmarks/Benchmarks.csproj [hunk 2]
```diff
@@ -0,0 +1,31 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+  <PropertyGroup>
+    <TargetFramework>netcoreapp1.1</TargetFramework>
+    <OutputType>Exe</OutputType>
+    <!-- NETStandard.Library version -->
```
Comment (mikeharder) [inline frameworks/CSharp/aspnetcore/Benchmarks/Benchmarks.csproj]: I would remove `NetStandardImplicitPackageVersion` and `RuntimeFrameworkVersion` unless most customer apps would set them.

File: frameworks/CSharp/aspnetcore/Benchmarks/Benchmarks.csproj [hunk 3]
```diff
@@ -0,0 +1,31 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+  <PropertyGroup>
+    <TargetFramework>netcoreapp1.1</TargetFramework>
```
Comment (mikeharder) [inline frameworks/CSharp/aspnetcore/Benchmarks/Benchmarks.csproj]: Let's move to `netcoreapp2.0` (2.0.0-preview1).

File: frameworks/CSharp/aspnetcore/Benchmarks/Benchmarks.csproj [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    <RuntimeFrameworkVersion>1.1.2</RuntimeFrameworkVersion>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <None Update="wwwroot\**\*" CopyToOutputDirectory="PreserveNewest" />
+    <None Include="*.json;Views\**\*" CopyToOutputDirectory="PreserveNewest" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Dapper" Version="1.50.2" />
```
Comment (mikeharder) [inline frameworks/CSharp/aspnetcore/Benchmarks/Benchmarks.csproj]: For third-party packages, please update to latest stable version on nuget.org.

For Microsoft packages, please update to latest shipped in 2.0.0-preview1.

File: toolset/setup/linux/languages/dotnetcore.sh [hunk 5]
```diff
@@ -6,7 +6,7 @@ sudo sh -c 'echo "deb [arch=amd64] https://apt-mo.trafficmanager.net/repos/dotne
 sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 417A0893
 sudo apt-get update
 
-sudo apt-get install dotnet-dev-1.0.0-preview2.1-003177 -y
+sudo apt-get install -y dotnet-dev-1.0.4
```
Comment (mikeharder) [inline toolset/setup/linux/languages/dotnetcore.sh]: Update to `2.0.0-preview1`.

**Comment by dougbu** [comment]
:up::date:

File: frameworks/CSharp/aspnetcore/Benchmarks/Benchmarks.csproj (hunk 4 above)
Comment (dougbu) [inline frameworks/CSharp/aspnetcore/Benchmarks/Benchmarks.csproj]: Modulo 1.1.3 -> 2.0.0-preview1, that was already the case.

**Comment by dougbu** [comment]
> We should also manually compare Benchmarks to FrameworkBenchmarks and manually port any changes to files used in both.

Done. Will shortly send out a Benchmarks PR containing the few things that looked better in this repo.

File: frameworks/CSharp/aspnetcore/Benchmarks/Data/EfDb.cs [hunk 6]
```diff
@@ -1,71 +1,92 @@
...
         private readonly ApplicationDbContext _dbContext;
+        private readonly bool _useBatchUpdate;
 
-        public EfDb(IRandom random, ApplicationDbContext dbContext)
+        public EfDb(IRandom random, ApplicationDbContext dbContext, IOptions<AppSettings> appSettings)
         {
             _random = random;
             _dbContext = dbContext;
-            _dbContext.ChangeTracker.QueryTrackingBehavior = QueryTrackingBehavior.NoTracking;
+            _useBatchUpdate = appSettings.Value.Database != DatabaseServer.PostgreSql;
```
Comment (divega) [inline frameworks/CSharp/aspnetcore/Benchmarks/Data/EfDb.cs]: @mikeharder do you recall why this was needed? If it was related to the deadlocks that affected once both PostgreSQL and SQL Server it should be removed. If there is still a reason to keep it, a comment would be a good idea.

File: frameworks/CSharp/aspnetcore/Benchmarks/Data/ApplicationDbContext.cs [hunk 7]
```diff
@@ -1,75 +1,21 @@
...
+    public sealed class ApplicationDbContext : DbContext
     {
-        private readonly AppSettings _appSettings;
-
-        public ApplicationDbContext(IOptions<AppSettings> appSettings)
+        public ApplicationDbContext(DbContextOptions options)
+            : base(options)
         {
-            _appSettings = appSettings.Value;
+            Database.AutoTransactionsEnabled = false;
```
Comment (divega) [inline frameworks/CSharp/aspnetcore/Benchmarks/Data/ApplicationDbContext.cs]: I am not sure if this provides any substantial perf. Ideally it shouldn't be needed.

File: frameworks/CSharp/aspnetcore/Benchmarks/Middleware/FortunesDapperMiddleware.cs [hunk 8]
```diff
@@ -16,21 +17,20 @@ public class FortunesDapperMiddleware
...
-            _db = db;
             _htmlEncoder = htmlEncoder;
         }
 
         public async Task Invoke(HttpContext httpContext)
         {
             if (httpContext.Request.Path.StartsWithSegments(_path, StringComparison.Ordinal))
             {
-                var rows = await _db.LoadFortunesRows();
+                var db = httpContext.RequestServices.GetService<DapperDb>();
```
Comment (divega) [inline frameworks/CSharp/aspnetcore/Benchmarks/Middleware/FortunesDapperMiddleware.cs]: Was this change necessary? I would prefer DI over SL if possible.

File: frameworks/CSharp/aspnetcore/Benchmarks/Startup.cs [hunk 9]
```diff
@@ -51,9 +51,23 @@ public void ConfigureServices(IServiceCollection services)
             // Common DB services
             services.AddSingleton<IRandom, DefaultRandom>();
             services.AddSingleton<ApplicationDbSeeder>();
-            services.AddEntityFrameworkSqlServer()
-                .AddDbContext<ApplicationDbContext>();
-            
+
+            services
+                .AddEntityFrameworkSqlServer()
+                .AddDbContextPool<ApplicationDbContext>((sp, ob) =>
```
Comment (divega) [inline frameworks/CSharp/aspnetcore/Benchmarks/Startup.cs]: I don't know how many times this delegate will be invoked. I would lean towards grabbing `appSettings.Database` and `appSettings.ConnectionString` early and conditionally calling `AddDbContextPool` with a simple delegate if possible.

File: frameworks/CSharp/aspnetcore/Benchmarks/Data/EfDb.cs (hunk 6 above)
Comment (dougbu) [inline frameworks/CSharp/aspnetcore/Benchmarks/Data/EfDb.cs]: FYI
- `BatchUpdate==false` part has been around what seems like forever
  - was refactored from `ApplicationDbContext` in this PR because that's how the Benchmarks code does it
  - controls whether `EfDb.LoadMultipleUpdatesRows()` calls `SaveChangesAsync()` inside or after its loop
- the `QueryTrackingBehavior` setting part is now in `ApplicationDbContext`

File: frameworks/CSharp/aspnetcore/Benchmarks/Middleware/FortunesDapperMiddleware.cs (hunk 8 above)
Comment (dougbu) [inline frameworks/CSharp/aspnetcore/Benchmarks/Middleware/FortunesDapperMiddleware.cs]: This middleware is a singleton while `DapperDb`, like `EfDb` and `RawDb`, is a scoped service. Shouldn't all `IDb` instances (and the `DBContext` instances they reference) be disposed at the end of a request?

File: frameworks/CSharp/aspnetcore/Benchmarks/Startup.cs (hunk 9 above)
Comment (dougbu) [inline frameworks/CSharp/aspnetcore/Benchmarks/Startup.cs]: This delegate is invoked once per application lifetime. Given that, does its simplicity matter?

File: frameworks/CSharp/aspnetcore/Benchmarks/Middleware/FortunesDapperMiddleware.cs (hunk 8 above)
Comment (divega) [inline frameworks/CSharp/aspnetcore/Benchmarks/Middleware/FortunesDapperMiddleware.cs]: Ok, if that is the case.

File: frameworks/CSharp/aspnetcore/Benchmarks/Startup.cs (hunk 9 above)
Comment (divega) [inline frameworks/CSharp/aspnetcore/Benchmarks/Startup.cs]: Thanks for checking. In that case it matters much less. 

If we care for code in the benchmark to look closer to what a regular app would do, then I prefer the simple if I see used elesewhere for variations on other dimensions.

File: frameworks/CSharp/aspnetcore/Benchmarks/Data/EfDb.cs (hunk 6 above)
Comment (divega) [inline frameworks/CSharp/aspnetcore/Benchmarks/Data/EfDb.cs]: I still believe it might be a leftover. Can try always enabling batching and see if it fails with PostgreSQL?

Comment (dougbu) [inline frameworks/CSharp/aspnetcore/Benchmarks/Data/EfDb.cs]: To confirm: leave the `QueryTrackingBehavior` alone?

Comment (divega) [inline frameworks/CSharp/aspnetcore/Benchmarks/Data/EfDb.cs]: I had nothing to say about `QueryTrackingBehavior`. Either setting is ok as long as on each individual query you are opting-in or out from tracking when it makes sense: queries performed without the purpose of saving changes on all entities returned can be faster if there is no tracking.

Comment (dougbu) [inline frameworks/CSharp/aspnetcore/Benchmarks/Data/EfDb.cs]: @divega :ok: good. `QueryTrackingBehavior` is fine because `LoadMultipleUpdatesRows()` is the only method that saves and its [query](https://github.com/aspnet/FrameworkBenchmarks/pull/3/files#diff-623cdd6ab5c4f009987652f7dbec928eR55) uses `AsTracking()`.

**Comment by dougbu** [comment]
:up::date: @divega / @mikeharder please have a quick look before I send this over to the TechEmpower repo.

**Comment by dougbu** [comment]
> did you have a chance to try @roji's recommendation to set Max Auto Prepare on the connection string &hellip; and getting rid of _useBatchUpdate

No. Filed aspnet/Performance/#236 because both involve tests of scenarios I haven't benchmarked.

**Comment by mikeharder** [comment]
Approved to submit PR to TechEmpower.

**Comment by dougbu** [comment]
Closing now that TechEmpower/FrameworkBenchmarks#2855 is open.

**Comment by dougbu** [comment]
Final commit was TechEmpower/FrameworkBenchmarks@70a2dac
